From d2599c34119ec1a2a15234c3650bfcde1d646c4b Mon Sep 17 00:00:00 2001
From: Smita Gaikwad <smitag@nvidia.com>
Date: Fri, 14 Sep 2012 07:12:45 -0400
Subject: [PATCH 2/3] [HTTP Live]Fetch another playlist when current fetch fails

In HTTP live streaming, while fetching a playlist having BW
less than or equal to estimated BW of n/w,selected
from the Master playlist fails, then try fetching another
playlist corresponding to the next lower BW till you reach
the playlist with lowest BW.

Bug 1045625

Change-Id: I48f7b257a476a96f437cd65f221687bd4efa59cf
---
 media/libstagefright/httplive/LiveSession.cpp |   19 ++++++++++++-------
 1 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/media/libstagefright/httplive/LiveSession.cpp b/media/libstagefright/httplive/LiveSession.cpp
index 93d6429..04360a1 100644
--- a/media/libstagefright/httplive/LiveSession.cpp
+++ b/media/libstagefright/httplive/LiveSession.cpp
@@ -506,11 +506,7 @@ rinse_repeat:
             || (ssize_t)bandwidthIndex != mPrevBandwidthIndex
             || (!mPlaylist->isComplete() && timeToRefreshPlaylist(nowUs))) {
         AString url;
-        if (mBandwidthItems.size() > 0) {
-            url = mBandwidthItems.editItemAt(bandwidthIndex).mURI;
-        } else {
-            url = mMasterURL;
-        }
+        url = mMasterURL;
 
         bool firstTime = (mPlaylist == NULL);
 
@@ -520,8 +516,17 @@ rinse_repeat:
             mPlaylist.clear();
         }
 
-        bool unchanged;
-        sp<M3UParser> playlist = fetchPlaylist(url.c_str(), &unchanged);
+        bool unchanged = false;
+        sp<M3UParser> playlist = NULL;
+        bandwidthIndex += 1;
+        while (playlist == NULL && bandwidthIndex > 0 && !unchanged) {
+            ALOGV("Fetch playlist with bandwidthIndex %d #############", bandwidthIndex);
+            if (mBandwidthItems.size() > 0) {
+                url = mBandwidthItems.editItemAt((bandwidthIndex-1)).mURI;
+            }
+            playlist = fetchPlaylist(url.c_str(), &unchanged);
+            --bandwidthIndex;//try for another playlist with lower BW if previous fetch failed
+        }
         if (playlist == NULL) {
             if (unchanged) {
                 // We succeeded in fetching the playlist, but it was
-- 
1.7.4.1

